### DISCUSSION TITLE
Optimize build time for Renovate PRs - CircleCI

### DISCUSSION THREAD

**Discussion by skjolber** [renovatebot/renovate#13883] [Suggest an Idea]
Hi, 

most of the apps in my workplace are within the Java ecosystem, with Maven and Gradle build systems. We use Github and Renovate bumps our dependencies both for private and public repositories. For building we use CircleCI + JFrog artifactory. 

For most builds, CircleCI is able to cache the app artifacts (dependencies). However, the default Maven / Gradle support in CircleCI has a flaw; caches are restored using a hash of the build files (pom / gradle build), thus our friend the Renovate bot effectively must start from scratch for even the smallest incremental changes to the build (i.e. all its PRs). This adds approximately 1 minute build time + transfer of the artifacts (we pay per GB) twice, once for the PR and (often) once for the PR merge. 

To work around this, we have created two CircleCI helpers, so-called orbs, for [Gradle](https://github.com/entur/gradle-orb) and [Maven](https://github.com/entur/maven-orb). These manage to restore the previous cache and also removes artifacts no longer in use.

What I am wondering if more Renovate user have observed similar problems on other build systems, and/or think this kind of build optimalization is worthwhile.

I know Github actions have a better, but not perfect, solution to this issue; instead of effectively wiping the cache like CircleCI, it will continue to add to a cache (which is updated more than once a week), risking cache load/save bloat (increased time before / after builds)

**Comment by rarkins** [comment]
Renovate no longer needs to run Gradle or Maven tools in order to update their package files, so I can't directly relate to your example. We used to use Gradle and it's still used in the deprecated "deep" extract.

**Reply by skjolber** [reply]
The above text refers to the build which is triggered once the PR is created (typically the PR is validated by a build). So it has nothing to do with how Renovate determines what to update.
